### ISSUE TITLE
Headers doesn’t export with exportToFile in the specific case 

### ISSUE THREAD

**Issue by stla** [handsontable/handsontable#4176]
Hello, 

Fiddle: https://jsfiddle.net/stla/hLw3zccj/6/

There's an empty cell in the first row, and `countSourceCols` returns `1` while there are two columns. Is it expected ?

**Comment by AMBudnik** [comment]
You're right, if we do not define the second column (here I have added the `content` key with an empty string as a value https://jsfiddle.net/handsoncode/9rtsv02j/) we won't get the 2 columns in the result of `countSourceCols`

**Comment by pnowak** [comment]
This error occurs because this method takes only the first element from the source data. I changed countSourceCols to include the largest number of columns.

PR - #4331.

**Comment by AMBudnik** [comment]
Still replicable https://jsfiddle.net/stla/hLw3zccj/6/ in v0.35.0

**Comment by cvjonesconvoy** [comment]
This issue, which I've seen in version 7.1, affects the exportFile plugin for CSV.  The columns exported for the CSV drop column labels for the column headers in the CSV file, since the number of columns in the table is undercounted.

**Comment by AMBudnik** [comment]
@cvjonesconvoy can you share a demo and/or a bug scenario?

**Comment by cvjonesconvoy** [comment]
Here's a jsFiddle that reproduces the issue: https://jsfiddle.net/pvr2qxye/1/

If you click the download button, here's the CSV file created:

```
,"S1",
T1,,10
T2,,20
T3,,30
T4,,40
T5,50,
T6,60,
T7,70,
```

The issue occurs when "columns" is defined as a function in the settings input to Handsontable. Because columns are defined as a function, getColHeader() returns just the first column, not two as are in the actual data.

**Comment by AMBudnik** [comment]
Hey @cvjonesconvoy 

here's what I see in the Fiddle (left) and in Excel (right)

![image](https://user-images.githubusercontent.com/10757813/60791031-1785f500-a163-11e9-8fc5-33f2a9be805a.png)

this seem to work as expected.

**Comment by cvjonesconvoy** [comment]
Apologies. I had forgotten to save the version with the bug. I have now saved it. Here's the link: https://jsfiddle.net/wg3nuyq9/6/

**Comment by AMBudnik** [comment]
It looks like the `S1` should be added to the first line of code

```js
{
    "S1": null,
    "S2": "10.000"
},
```

**Comment by cvjonesconvoy** [comment]
If I'm understanding the suggestion, the first row of the data set has to have some value for each column in order for the code to count the correct number of columns. Although that's a workaround that can likely work, I'm surprised that it's necessary. 

In particular, the problem only occurs when "columns" is a function.  Here's the "columns" function in the example:
```
"columns" : function(colNo) {
			  return {
			    "data": colCodes[colNo]
			  }
			}, 
```
Where colCodes is:
```
["S1","S2"]
```

In conclusion, why is it so difficult for handsontable to count the number of columns, especially since it is capable of displaying the right number of columns even with "sparse" data input?

**Comment by wojciechczerniak** [comment]
Because it is difficult.

You have combined two ambiguity factors: `data` as an object which doesn't have all properties (columns) defined and `columns` as a function which needs to be called for each column. But there is no way to tell how many columns are there.

**Comment by cvjonesconvoy** [comment]
A couple of questions, then.

1. The settings include colHeaders, which is a list of size 2. Could that be a useful source of column size?
2. The sparse table renders beautifully as HTML. Can't that rendering approach also then be used to generate the CSV? Since CSV is actually simpler, such an approach, I think, could work well.
3. Does the documentation for the exportFile plugin describe the need to fill out the data completely in order to generate a CSV file robustly?

Thanks,
Chris

**Comment by wojciechczerniak** [comment]
Good questions. `colHeaders` can be a function too, so it would end up as another ambiguity factor. Of course CSV should match the view. This is a bug that we're reading columns count in both places differently. I've checked this and rendering fallbacks to `dataSchema` to count columns. CSV should too. 

IMO ExportFile docs shouldn't have this kind of description. I'm just explaining how it works so you can workaround the bug until it's fixed.

**Comment by aninde** [comment]
In v. `8.0.0-rc1` bug is still replicable. 

**Demo**
*Stla’s Case*
v. `0.31.2` https://jsfiddle.net/aninde/pbaq6cft/ 
v `8.0.0-rc1` - forked original demo https://jsfiddle.net/aninde/rcj0fbz2/ 

*Importing to the CSV case*
v. `8.0.0-rc1` https://jsfiddle.net/aninde/21nh8pbf/

_Left image- Handsontable, right image- imported data_ 
![Screenshot 2019-09-29 at 09 26 09](https://user-images.githubusercontent.com/30186818/65828807-292cf380-e29f-11e9-85ba-3f41f8ee8c48.png)

**Comment by wojciechczerniak** [comment]
@aninde Works ok in v8-beta.2?  https://jsfiddle.net/c4a8xhs6/ 
![Screenshot 2020-07-02 at 00 39 27](https://user-images.githubusercontent.com/4718998/86297809-912cb380-bbfc-11ea-839e-1a8e72253425.png)

**Comment by AMBudnik** [comment]
@wojciechczerniak actually, I get return '2' even in the v 0.31.2 https://jsfiddle.net/4jzgs3on/1/ 
Tested on macOS Catalina and Chrome 83

**Comment by aninde** [comment]
After casting to the original demo version `8.0.0-beta2-rev18`  incorrect 1 is still returned by `countSourceCols()`  https://jsfiddle.net/aninde/1h025nz8/ on Firefox and Chrome.

**Comment by aninde** [comment]
@AMBudnik you added key `content: null` in your demo for `data`. This makes it a really different use case than the one originally reported. The question is, should we then consider this issue when we have not added the  key`content` to the `data`? 
![Screenshot 2020-07-03 at 13 02 12](https://user-images.githubusercontent.com/30186818/86463459-d56ba100-bd2d-11ea-99df-0a2a18398892.png)

CC: @wojciechczerniak

**Comment by wojciechczerniak** [comment]
@AMBudnik Right! I took the last repro only and was looking at the @cvjonesconvoy case.

To the original poster and main case:
> Is it expected?

Yes, it is.

`countSourceCols()` will take value either from `schema`, `columns` or the FIRST row from the data set. As mentioned in the comments above I strongly advise specifying one of those to avoid ambiguity. Or use a constructor object for data row to be sure that every row is exactly the same.

Checking each row to choose the max length is not an option. This would kill the data load performance.

**Comment by aninde** [comment]
To solve the first case - incorrect number of columns with countSourceCols() I suggest after @AMBudnik and @wojciechczerniak  adding key `content: null` to the `data`
```
data: [{
content: null
}]
```

Coming back to the second use case with Importing to the CSV case. 
Behavior does not change in `8.0.0-beta2-rev18` https://jsfiddle.net/aninde/e4ajuhzk/ without introducing @wojciechczerniak's recommendation scountSourceCols() still return 1 in the same exact case as reported here https://github.com/handsontable/handsontable/issues/4176#issuecomment-509191997 because `exportToFile` works differently.
![Screenshot 2020-07-03 at 13 30 14](https://user-images.githubusercontent.com/30186818/86465337-b53de100-bd31-11ea-8ec1-54d631acaa1c.png).

Our documentation should describe the correct usage, so it will be reported for completion in it.

**Comment by aninde** [comment]
Case 1 is reported here https://github.com/handsontable/docs/issues/120. Due to the fact that we have determined its solution, and case 2 concerns rendering during export and it is fixed, I change the title.

**Comment by AMBudnik** [comment]
Hi @stla @cvjonesconvoy 

We've just released the 8.0.0-beta2 version that solves the following issue. I would really appreciate your feedback. I hope that it works for you as well.

Here’s a full list of changes [https://github.com/handsontable/handsontable/releases/tag/8.0.0-beta.2](https://slack-redir.net/link?url=https%3A%2F%2Fgithub.com%2Fhandsontable%2Fhandsontable%2Freleases%2Ftag%2F8.0.0-beta.2)

NPM [https://www.npmjs.com/package/handsontable/v/8.0.0-beta.2](https://slack-redir.net/link?url=https%3A%2F%2Fwww.npmjs.com%2Fpackage%2Fhandsontable%2Fv%2F8.0.0-beta.2)

**Comment by AMBudnik** [comment]
Hi @stla 

I’m glad to announce that Handsontable v8.0.0 is live. This issue should already be solved. I highly encourage you to test the new version and leave some feedback.

CDN https://cdn.jsdelivr.net/gh/handsontable/handsontable@latest/dist/handsontable.full.css
https://cdn.jsdelivr.net/gh/handsontable/handsontable@latest/dist/handsontable.full.js
NPM npm i handsontable
Release Notes https://github.com/handsontable/handsontable/releases
